File: forge-cli/tools/browser/snapshot.go
package browser

import (
	"context"
	_ "embed"
	"encoding/json"
	"errors"
	"fmt"
	"time"

	"github.com/chromedp/cdproto/input"
	"github.com/chromedp/chromedp"
	"github.com/chromedp/chromedp/kb"
)

//go:embed snapshot.js
var snapshotJS string

//go:embed extract.js
var extractJS string

// ErrStale is returned when an interaction references a generation older than
// the current page state (the page navigated or was re-snapshotted since the
// LLM last saw it). Tools catch it and return a fresh digest so the model
// recovers in one turn.
var ErrStale = errors.New("stale element index: the page changed since the last snapshot")

// elementInfo mirrors one entry of snapshot.js's els array.
type elementInfo struct {
	Index     int      `json:"i"`
	Tag       string   `json:"tag"`
	Role      string   `json:"role"`
	Name      string   `json:"name"`
	Href      string   `json:"href,omitempty"`
	InputType string   `json:"inputType,omitempty"`
	Checked   *bool    `json:"checked,omitempty"`
	Options   []string `json:"options,omitempty"`
	Value     string   `json:"value,omitempty"`
	Protected bool     `json:"protected"`
	CX        float64  `json:"cx"`
	CY        float64  `json:"cy"`
}

// pageSnapshot mirrors snapshot.js's return value.
type pageSnapshot struct {
	URL      string        `json:"url"`
	Title    string        `json:"title"`
	Gen      int64         `json:"gen"`
	Els      []elementInfo `json:"els"`
	TotalEls int           `json:"totalEls"`
	Text     string        `json:"text"`
	TextLen  int           `json:"textLen"`
}

const (
	defaultMaxEls  = 100
	defaultMaxText = 1200
	settleDelay    = 300 * time.Millisecond
)

type point struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

// snapshotAction captures a fresh snapshot into snap under a new generation.
// Must run inside m.run (m.mu held).
func (m *Manager) snapshotAction(snap *pageSnapshot, maxEls int) chromedp.Action {
	return chromedp.ActionFunc(func(ctx context.Context) error {
		m.gen++
		opts, err := json.Marshal(map[string]any{
			"maxEls":  maxEls,
			"maxText": defaultMaxText,
			"gen":     m.gen,
		})
		if err != nil {
			return err
		}
		expr := "(" + snapshotJS + ")(" + string(opts) + ")"
		if err := chromedp.Evaluate(expr, snap).Do(ctx); err != nil {
			return fmt.Errorf("snapshot: %w", err)
		}
		return nil
	})
}

// checkGenAction verifies the page still matches the generation the LLM acted
// on. A missing __forge_gen (page navigated, window state wiped) or a mismatch
// yields ErrStale.
func checkGenAction(gen int64) chromedp.Action {
	return chromedp.ActionFunc(func(ctx context.Context) error {
		var cur int64
		expr := "typeof window.__forge_gen === 'number' ? window.__forge_gen : -1"
		if err := chromedp.Evaluate(expr, &cur).Do(ctx); err != nil {
			return fmt.Errorf("generation check: %w", err)
		}
		if cur != gen {
			return ErrStale
		}
		return nil
	})
}

// settleAction waits briefly for click/fill side effects (navigation, SPA
// re-render) to land: fixed delay, then poll readyState out of 'loading'.
func settleAction() chromedp.Action {
	return chromedp.ActionFunc(func(ctx context.Context) error {
		if err := chromedp.Sleep(settleDelay).Do(ctx); err != nil {
			return err
		}
		deadline := time.Now().Add(5 * time.Second)
		for time.Now().Before(deadline) {
			var state string
			if err := chromedp.Evaluate("document.readyState", &state).Do(ctx); err == nil && state != "loading" {
				return nil
			}
			if err := chromedp.Sleep(100 * time.Millisecond).Do(ctx); err != nil {
				return err
			}
		}
		return nil // settled or not, let the snapshot show reality
	})
}

// Navigate loads url and returns a fresh snapshot.
func (m *Manager) Navigate(url string, waitMS int, maxEls int) (pageSnapshot, error) {
	if maxEls <= 0 {
		maxEls = defaultMaxEls
	}
	extraWait := time.Duration(waitMS) * time.Millisecond
	if extraWait < 0 {
		extraWait = 0
	}
	if extraWait > 15*time.Second {
		extraWait = 15 * time.Second
	}
	var snap pageSnapshot
	err := m.run(m.cfg.NavTimeout,
		chromedp.Navigate(url),
		chromedp.Sleep(extraWait),
		settleAction(),
		m.snapshotAction(&snap, maxEls),
	)
	return snap, err
}

// Snapshot re-reads the current page (optionally scrolling first).
func (m *Manager) Snapshot(maxEls int, scrollToIndex int, scrollPages float64) (pageSnapshot, error) {
	if maxEls <= 0 {
		maxEls = defaultMaxEls
	}
	var actions []chromedp.Action
	if scrollToIndex >= 0 {
		actions = append(actions, chromedp.ActionFunc(func(ctx context.Context) error {
			var pt *point
			expr := fmt.Sprintf("window.__forge_center ? window.__forge_center(%d) : null", scrollToIndex)
			return chromedp.Evaluate(expr, &pt).Do(ctx) // scrollIntoView side effect; nil pt is fine
		}))
	} else if scrollPages != 0 {
		actions = append(actions, chromedp.ActionFunc(func(ctx context.Context) error {
			expr := fmt.Sprintf("window.scrollBy(0, window.innerHeight * %g); true", scrollPages)
			var ok bool
			return chromedp.Evaluate(expr, &ok).Do(ctx)
		}))
	}
	var snap pageSnapshot
	actions = append(actions, settleAction(), m.snapshotAction(&snap, maxEls))
	err := m.run(m.cfg.ActionTimeout, actions...)
	return snap, err
}

// Click resolves index to fresh viewport coordinates and dispatches a trusted
// CDP mouse click, then returns the post-action snapshot.
func (m *Manager) Click(index int, gen int64, maxEls int) (pageSnapshot, error) {
	if maxEls <= 0 {
		maxEls = defaultMaxEls
	}
	var snap pageSnapshot
	err := m.run(m.cfg.ActionTimeout,
		checkGenAction(gen),
		chromedp.ActionFunc(func(ctx context.Context) error {
			var pt *point
			expr := fmt.Sprintf("window.__forge_center(%d)", index)
			if err := chromedp.Evaluate(expr, &pt).Do(ctx); err != nil {
				return fmt.Errorf("resolve element %d: %w", index, err)
			}
			if pt == nil {
				return fmt.Errorf("%w (element [%d] no longer attached)", ErrStale, index)
			}
			return chromedp.MouseClickXY(pt.X, pt.Y).Do(ctx)
		}),
		settleAction(),
		m.snapshotAction(&snap, maxEls),
	)
	return snap, err
}

// Fill types text into the element at index: focus via trusted click, select
// existing content, replace it with CDP Input.insertText (fires native input
// events, so React/Vue controlled inputs see the change), then dispatch
// change. Select elements pick the matching option instead. Protected fields
// (password/payment autocomplete) are refused unless allowSensitive.
func (m *Manager) Fill(index int, text string, gen int64, submit bool, allowSensitive bool, maxEls int) (pageSnapshot, error) {
	if maxEls <= 0 {
		maxEls = defaultMaxEls
	}
	var snap pageSnapshot
	err := m.run(m.cfg.ActionTimeout,
		checkGenAction(gen),
		chromedp.ActionFunc(func(ctx context.Context) error {
			// Re-check protection in the live DOM, not a cached snapshot: a
			// hostile page could morph a text field into a password field
			// after the snapshot was taken.
			if !allowSensitive {
				var prot bool
				expr := fmt.Sprintf("window.__forge_protected(%d)", index)
				if err := chromedp.Evaluate(expr, &prot).Do(ctx); err != nil {
					return fmt.Errorf("protection check: %w", err)
				}
				if prot {
					return fmt.Errorf("field [%d] is fill-protected (password or payment input); the skill must opt in via guardrails browser.allow_sensitive_fill", index)
				}
			}

			var isSelect bool
			if err := chromedp.Evaluate(fmt.Sprintf(
				"(function(){var el=window.__forge_els[%d]; return !!el && el.tagName.toLowerCase()==='select';})()", index),
				&isSelect).Do(ctx); err != nil {
				return fmt.Errorf("resolve element %d: %w", index, err)
			}
			if isSelect {
				var ok bool
				expr := fmt.Sprintf("window.__forge_select_option(%d, %s)", index, mustJSON(text))
				if err := chromedp.Evaluate(expr, &ok).Do(ctx); err != nil {
					return fmt.Errorf("select option: %w", err)
				}
				if !ok {
					return fmt.Errorf("no option matching %q in select [%d]; see its options in the digest", text, index)
				}
				return nil
			}

			// Focus with a real click so focus/blur handlers fire.
			var pt *point
			if err := chromedp.Evaluate(fmt.Sprintf("window.__forge_center(%d)", index), &pt).Do(ctx); err != nil {
				return fmt.Errorf("resolve element %d: %w", index, err)
			}
			if pt == nil {
				return fmt.Errorf("%w (element [%d] no longer attached)", ErrStale, index)
			}
			if err := chromedp.MouseClickXY(pt.X, pt.Y).Do(ctx); err != nil {
				return err
			}
			var selected bool
			if err := chromedp.Evaluate(fmt.Sprintf("window.__forge_select_all(%d)", index), &selected).Do(ctx); err != nil {
				return fmt.Errorf("select existing content: %w", err)
			}
			if err := input.InsertText(text).Do(ctx); err != nil {
				return fmt.Errorf("insert text: %w", err)
			}
			var changed bool
			if err := chromedp.Evaluate(fmt.Sprintf("window.__forge_dispatch_change(%d)", index), &changed).Do(ctx); err != nil {
				return fmt.Errorf("dispatch change: %w", err)
			}
			if submit {
				return chromedp.KeyEvent(kb.Enter).Do(ctx)
			}
			return nil
		}),
		settleAction(),
		m.snapshotAction(&snap, maxEls),
	)
	return snap, err
}

// Extract returns the page content in the requested mode ("text", "links",
// "html") plus the current URL. Pagination happens in the tool layer.
func (m *Manager) Extract(mode, selector string) (content string, url string, err error) {
	err = m.run(m.cfg.ActionTimeout,
		chromedp.Location(&url),
		chromedp.ActionFunc(func(ctx context.Context) error {
			expr := "(" + extractJS + ")(" + mustJSON(mode) + ", " + mustJSON(selector) + ")"
			return chromedp.Evaluate(expr, &content).Do(ctx)
		}),
	)
	return content, url, err
}

// Screenshot captures the viewport (or full page) as PNG bytes.
func (m *Manager) Screenshot(fullPage bool) ([]byte, error) {
	var buf []byte
	var action chromedp.Action
	if fullPage {
		action = chromedp.FullScreenshot(&buf, 100)
	} else {
		action = chromedp.CaptureScreenshot(&buf)
	}
	if err := m.run(m.cfg.ActionTimeout, action); err != nil {
		return nil, err
	}
	return buf, nil
}

// mustJSON marshals a string for safe embedding in a JS expression.
func mustJSON(s string) string {
	b, err := json.Marshal(s)
	if err != nil {
		// strings always marshal; keep the assertion visible
		panic(fmt.Sprintf("browser: marshal string: %v", err))
	}
	return string(b)
}

// truncate returns s capped at max runes-ish (byte-safe on ASCII, close
// enough for caps) with a marker.
func truncate(s string, max int) string {
	if len(s) <= max {
		return s
	}
	cut := max
	for cut > 0 && !utf8Start(s[cut]) {
		cut--
	}
	return s[:cut] + "…"
}

func utf8Start(b byte) bool { return b&0xC0 != 0x80 }

File: forge-cli/tools/browser/snapshot.js
// snapshot.js — indexed interactive-element digest builder.
//
// Evaluated by the Go side as `(function(opts){...})({maxEls, maxText, gen})`.
// Walks the DOM (including shadow roots and same-origin iframes), filters to
// visible interactive elements, stores live element references in
// window.__forge_els so tools can act by index, and returns a compact
// serializable snapshot. This file is a bare function expression, not a
// standalone script.
function (opts) {
	opts = opts || {};
	var maxEls = opts.maxEls || 100;
	var maxText = opts.maxText || 1200;
	var gen = opts.gen || 0;

	var els = [];
	var infos = [];

	var SKIP_TAGS = { script: 1, style: 1, noscript: 1, template: 1, meta: 1, link: 1, head: 1 };
	var INTERACTIVE_ROLES = {
		button: 1, link: 1, tab: 1, menuitem: 1, checkbox: 1, radio: 1,
		combobox: 1, switch: 1, option: 1, textbox: 1, searchbox: 1, slider: 1
	};
	var PROTECTED_AUTOCOMPLETE = /^cc-|^(current|new)-password$/;

	function isVisible(el) {
		try {
			if (el.checkVisibility && !el.checkVisibility()) return false;
		} catch (e) { /* fall through to rect check */ }
		var r = el.getBoundingClientRect();
		return r.width > 0 && r.height > 0;
	}

	function isInteractive(el) {
		var tag = el.tagName.toLowerCase();
		if (tag === 'a') return el.hasAttribute('href');
		if (tag === 'button' || tag === 'select' || tag === 'textarea' || tag === 'summary') return true;
		if (tag === 'input') return (el.getAttribute('type') || '').toLowerCase() !== 'hidden';
		var ce = el.getAttribute('contenteditable');
		if (ce !== null && ce !== 'false') return true;
		if (el.hasAttribute('onclick')) return true;
		return !!INTERACTIVE_ROLES[(el.getAttribute('role') || '').toLowerCase()];
	}

	function isProtected(el) {
		if (el.tagName.toLowerCase() !== 'input') return false;
		if ((el.getAttribute('type') || '').toLowerCase() === 'password') return true;
		return PROTECTED_AUTOCOMPLETE.test((el.getAttribute('autocomplete') || '').toLowerCase());
	}

	function clean(s, cap) {
		s = (s || '').replace(/\s+/g, ' ').trim();
		if (s.length > cap) s = s.slice(0, cap - 3) + '...';
		return s;
	}

	function accessibleName(el) {
		var name = el.getAttribute('aria-label') || '';
		if (!name) {
			var lb = el.getAttribute('aria-labelledby');
			if (lb) {
				name = lb.split(/\s+/).map(function (id) {
					var n = el.ownerDocument.getElementById(id);
					return n ? (n.innerText || n.textContent || '') : '';
				}).join(' ');
			}
		}
		if (!name) {
			var tag = el.tagName.toLowerCase();
			if (tag === 'input' || tag === 'textarea') {
				name = (el.labels && el.labels.length ? el.labels[0].innerText : '') ||
					el.getAttribute('placeholder') || el.getAttribute('name') || '';
			} else if (tag === 'select') {
				name = (el.labels && el.labels.length ? el.labels[0].innerText : '') ||
					el.getAttribute('name') || '';
			} else if (tag === 'img') {
				name = el.getAttribute('alt') || '';
			} else {
				name = el.innerText || el.textContent || '';
			}
		}
		if (!name) name = el.getAttribute('title') || el.getAttribute('value') || '';
		return clean(name, 80);
	}

	function roleOf(el) {
		var tag = el.tagName.toLowerCase();
		var role = (el.getAttribute('role') || '').toLowerCase();
		if (tag === 'a') return 'link';
		if (tag === 'input') return 'input';
		if (tag === 'button' || tag === 'select' || tag === 'textarea' || tag === 'summary') return tag;
		var ce = el.getAttribute('contenteditable');
		if (ce !== null && ce !== 'false') return 'editable';
		return role || tag;
	}

	// viewport-relative center of an element, accounting for same-origin
	// iframe nesting (rects inside a frame are relative to that frame).
	function viewportCenter(el) {
		var r = el.getBoundingClientRect();
		var x = r.left + r.width / 2;
		var y = r.top + r.height / 2;
		try {
			var win = el.ownerDocument.defaultView;
			while (win && win.frameElement) {
				var fr = win.frameElement.getBoundingClientRect();
				x += fr.left;
				y += fr.top;
				win = win.parent;
			}
		} catch (e) { /* cross-origin parent: keep local coords */ }
		return { x: x, y: y };
	}

	function record(el) {
		var idx = els.length;
		els.push(el);
		var c = viewportCenter(el);
		var info = {
			i: idx,
			tag: el.tagName.toLowerCase(),
			role: roleOf(el),
			name: accessibleName(el),
			protected: isProtected(el),
			cx: Math.round(c.x * 10) / 10,
			cy: Math.round(c.y * 10) / 10
		};
		if (info.tag === 'a') {
			info.href = clean(el.getAttribute('href') || '', 120);
		}
		if (info.tag === 'input') {
			info.inputType = (el.getAttribute('type') || 'text').toLowerCase();
			if (info.inputType === 'checkbox' || info.inputType === 'radio') {
				info.checked = !!el.checked;
			}
		}
		if (info.tag === 'select') {
			var options = [];
			for (var i = 0; i < el.options.length && options.length < 10; i++) {
				options.push(clean(el.options[i].text, 40));
			}
			if (el.options.length > 10) options.push('… +' + (el.options.length - 10) + ' more');
			info.options = options;
			if (el.selectedOptions && el.selectedOptions.length) {
				info.value = clean(el.selectedOptions[0].text, 40);
			}
		}
		infos.push(info);
	}

	function visit(el) {
		if (!el || el.nodeType !== 1) return;
		var tag = el.tagName.toLowerCase();
		if (SKIP_TAGS[tag]) return;

		if (tag === 'iframe' || tag === 'frame') {
			try {
				var doc = el.contentDocument;
				if (doc && doc.body) visit(doc.body);
			} catch (e) { /* cross-origin iframe: skip */ }
			return;
		}

		if (isInteractive(el) && isVisible(el)) record(el);

		if (el.shadowRoot) {
			for (var s = 0; s < el.shadowRoot.children.length; s++) visit(el.shadowRoot.children[s]);
		}
		for (var i = 0; i < el.children.length; i++) visit(el.children[i]);
	}

	visit(document.body || document.documentElement);

	// Side effects: the index → element map and helpers the interaction
	// tools use. A navigation wipes these, which is exactly the staleness
	// signal (window.__forge_gen becomes undefined).
	window.__forge_els = els;
	window.__forge_gen = gen;
	window.__forge_center = function (i) {
		var el = window.__forge_els[i];
		if (!el || !el.isConnected) return null;
		try { el.scrollIntoView({ block: 'center', inline: 'nearest', behavior: 'instant' }); } catch (e) { el.scrollIntoView(); }
		return viewportCenter(el);
	};
	window.__forge_protected = function (i) {
		var el = window.__forge_els[i];
		return !el || !el.isConnected || isProtected(el);
	};
	window.__forge_select_all = function (i) {
		var el = window.__forge_els[i];
		if (!el || !el.isConnected) return false;
		el.focus();
		if (el.select) {
			el.select();
		} else {
			var range = el.ownerDocument.createRange();
			range.selectNodeContents(el);
			var sel = el.ownerDocument.defaultView.getSelection();
			sel.removeAllRanges();
			sel.addRange(range);
		}
		return true;
	};
	window.__forge_select_option = function (i, label) {
		var el = window.__forge_els[i];
		if (!el || el.tagName.toLowerCase() !== 'select') return false;
		for (var j = 0; j < el.options.length; j++) {
			var o = el.options[j];
			if (o.text.trim() === label || o.value === label) {
				el.selectedIndex = j;
				el.dispatchEvent(new Event('input', { bubbles: true }));
				el.dispatchEvent(new Event('change', { bubbles: true }));
				return true;
			}
		}
		return false;
	};
	window.__forge_dispatch_change = function (i) {
		var el = window.__forge_els[i];
		if (!el || !el.isConnected) return false;
		el.dispatchEvent(new Event('change', { bubbles: true }));
		return true;
	};

	var bodyText = clean(document.body ? document.body.innerText : '', 10000000);
	return {
		url: location.href,
		title: document.title || '',
		gen: gen,
		els: infos.slice(0, maxEls),
		totalEls: infos.length,
		text: bodyText.slice(0, maxText),
		textLen: bodyText.length
	};
}
